internal/pkg/manifest/union.go

```diff
@@ -62,48 +66,64 @@ func (t Union[_, _]) IsAdvanced() bool {
 }
 
 // UnmarshalYAML decodes value into either type Basic or Advanced, and stores that value
-// in t. value is first decoded into type Basic, and t will hold type Basic if:
-//   - There was no error decoding value into type Basic
-//   - Basic.IsZero() returns false OR Basic does not support IsZero()
+// in t. value is first decoded into type Basic, and t will hold type Basic if
+// (1) There was no error decoding value into type Basic and
+// (2) Basic.IsZero() returns false OR Basic is not zero via reflection.
 //
 // If Basic didn't meet the above criteria, then value is decoded into type Advanced.
 // t will hold type Advanced if Advanced meets the same conditions that were required for type Basic.
 //
 // If value fails to decode into either type, or both types are zero after
 // decoding, t will not hold any value.
 func (t *Union[Basic, Advanced]) UnmarshalYAML(value *yaml.Node) error {
-	// reset struct
-	var simple Basic
-	var advanced Advanced
-	t.isBasic, t.Basic = false, simple
-	t.isAdvanced, t.Advanced = false, advanced
+	// convert value to []byte so we can use KnownFields()
+	// see https://github.com/go-yaml/yaml/issues/460
+	// side effect: TypeError line numbers are reset.
+	b, err := yaml.Marshal(value)
+	if err != nil {
+		return err
+	}
 
-	sErr := value.Decode(&simple)
-	if sErr == nil && !isZeroerAndZero(simple) {
-		t.isBasic, t.Basic = true, simple
+	var basic Basic
+	dec := yaml.NewDecoder(bytes.NewReader(b))
+	dec.KnownFields(true)
+	bErr := dec.Decode(&basic)
+	if bErr == nil && !isZero(basic) {
+		t.SetBasic(basic)
 		return nil
 	}
 
-	aErr := value.Decode(&advanced)
-	if aErr == nil && !isZeroerAndZero(advanced) {
-		t.isAdvanced, t.Advanced = true, advanced
+	var advanced Advanced
+	dec = yaml.NewDecoder(bytes.NewReader(b))
+	dec.KnownFields(true)
+	aErr := dec.Decode(&advanced)
+	if aErr == nil && !isZero(advanced) {
+		t.SetAdvanced(advanced)
 		return nil
 	}
 
-	if sErr != nil {
-		return sErr
+	switch {
+	case bErr == nil && aErr == nil:
+		return nil
+	case bErr != nil && aErr == nil:
+		return bErr
+	case aErr != nil && bErr == nil:
+		return aErr
 	}
-	return aErr
+
+	// multiline error because yaml.TypeError (which this likely is)
+	// is already a multiline error
+	return fmt.Errorf("%T: %s\n%T: %s", t.Basic, bErr, t.Advanced, aErr)
 }
 
-// isZeroerAndZero returns true if v is a yaml.Zeroer
-// _and_ is zero, as determined by v.IsZero().
-func isZeroerAndZero(v any) bool {
-	z, ok := v.(yaml.IsZeroer)
-	if !ok {
-		return false
+// isZero returns true if:
+//   - v is a yaml.Zeroer and IsZero().
+//   - v is not a yaml.Zeroer and determined to be zero via reflection.
+func isZero(v any) bool {
+	if z, ok := v.(yaml.IsZeroer); ok {
+		return z.IsZero()
 	}
-	return z.IsZero()
+	return reflect.ValueOf(v).IsZero()
```

Comment:
Could this panic 🤔

Reply:
Do you know if a pointer type always implements `yaml.IsZerorer`? Are we falling through `reflect` so that we don't have to implement `IsZero` for every single custom `struct`? If so, I wonder if it actually makes sense to require the developers to implement `IsZero` themselves?

Reply:
it is possible if you called `isZero(nil)` - but since it's a private function and only used here i think it's probably fine - happy to add some extra logic if you think we should though.

Reply:
> Do you know if a pointer type always implements yaml.IsZerorer

No, only if there is a function like `(c *Config) IsZero() bool`

> Are we falling through reflect so that we don't have to implement IsZero for every single custom struct

Yes, but also for comparison to builtin types. i.e., you can use `Union[string, []string]` without having to create a custom `string` type that has `IsZero()`

Reply:
Sounds good!

I was trying to  think of a scenario where a `struct`'s zero value determined by `reflect` isn't really the zero value that we agree with. If this is potentially possible, then it might make sense to figure out a way to force developers to implement `IsZero` instead of automatically falling through to `reflect`. But I don't think I have thought of any, so we  should be good here.